browser/ephemeral_storage/ephemeral_storage_tab_helper.cc

```diff
@@ -0,0 +1,183 @@
+  auto* partition =
+      BrowserContext::GetStoragePartition(browser_context, site_instance.get());
+
+  std::string local_partition_id = domain + "/ephemeral-local-storage";
+  content::SessionStorageNamespaceMap::const_iterator it =
+      local_storage_namespace_map().find(local_partition_id);
+  if (it == local_storage_namespace_map().end()) {
+    auto local_storage_namespace =
+        content::CreateSessionStorageNamespace(partition, local_partition_id);
+    local_storage_namespace_map()[local_partition_id] =
+        std::move(local_storage_namespace);
+  }
+
+  std::string session_partition_id =
+      content::GetSessionStorageNamespaceId(web_contents()) +
+      "/ephemeral-session-storage";
+  it = session_storage_namespace_map().find(session_partition_id);
+  if (it == session_storage_namespace_map().end()) {
+    auto session_storage_namespace =
+        content::CreateSessionStorageNamespace(partition, session_partition_id);
+    session_storage_namespace_map()[session_partition_id] =
+        std::move(session_storage_namespace);
+  }
+}
+
+void EphemeralStorageTabHelper::WebContentsDestroyed() {
+  ClearEphemeralStorageIfNecessary();
+}
+
+bool EphemeralStorageTabHelper::IsDifferentWebContentsWithMatchingStorageDomain(
+    WebContents* other_web_contents,
+    const std::string& storage_domain) {
+  DCHECK(other_web_contents);
+  if (web_contents() == other_web_contents)
+    return false;
+
+  std::string other_storage_domain =
+      URLToStorageDomain(other_web_contents->GetLastCommittedURL());
+  return storage_domain == other_storage_domain;
+}
+
+bool EphemeralStorageTabHelper::IsAnotherTabOpenWithStorageDomain(
+    const std::string& storage_domain) {
+#if defined(OS_ANDROID)
+  for (auto it = TabModelList::begin(); it != TabModelList::end(); ++it) {
+    TabModel* tab_model = *it;
+    if (tab_model->GetProfile() == profile) {
+      for (int i = 0; i < tab_model->GetTabCount(); i++) {
+        TabAndroid* tab = tab_model->GetTabAt(i);
+        WebContents* web_contents = tab->web_contents();
+        if (IsDifferentWebContentsWithMatchingStorageDomain(web_contents,
+                                                            storage_domain))
+          return true;
+      }
+    }
+  }
+#else
+  for (Browser* browser : *BrowserList::GetInstance()) {
+    if (browser->profile() == web_contents()->GetBrowserContext()) {
+      TabStripModel* tab_strip = browser->tab_strip_model();
+      for (int i = 0; i < tab_strip->count(); ++i) {
+        WebContents* web_contents = tab_strip->GetWebContentsAt(i);
+        if (IsDifferentWebContentsWithMatchingStorageDomain(web_contents,
+                                                            storage_domain))
+          return true;
+      }
+    }
+  }
+#endif
+
+  return false;
+}
+
+void EphemeralStorageTabHelper::ClearEphemeralStorageIfNecessary() {
+  if (!base::FeatureList::IsEnabled(blink::features::kBraveEphemeralStorage)) {
```

Comment:
maybe we can just skip creation of the helper at all?

Reply:
At some point, this will probably be a feature that can be toggled on and of at runtime. Does it make sense to structure the code as if that were the case?

Reply:
probably yes, but this could be tricky, for instance disabling the feature at runtime will block `ClearEphemeralStorageIfNecessary()`, so existed storage would remain dirty. So I'd rather keep things simple, but I'm not a feature owner

Reply:
Okay. I'll go with the approach you suggest.